### PR TITLE
restructuring parachain commitment relayer into isolated worker

### PR DESCRIPTION
**Pull request by musnit**
- e2e tests all pass
- all functionality from existing parachain chain and ethereum chain related to relaying parachain commitments has been removed
- basically ready for review, though may want to get @Rizziepit 's worker pool PR in first

### CODE REVIEW

File: relayer/chain/ethereum/connection.go [hunk 1]
```diff
@@ -57,3 +57,7 @@ func (co *Connection) Close() {
 func (co *Connection) GetClient() *ethclient.Client {
 	return co.client
 }
+
+func (co *Connection) GetKP() *secp256k1.Keypair {
```
Comment (vgeddes) [inline relayer/chain/ethereum/connection.go:61]: No need to prepend `Get` to the getter name, if one follows golang stylistic guidance.

File: relayer/chain/parachain/connection.go [hunk 2]
```diff
@@ -22,6 +22,10 @@ type Connection struct {
 	log         *logrus.Entry
 }
 
+func (co *Connection) GetAPI() *gsrpc.SubstrateAPI {
```
Comment (vgeddes) [inline relayer/chain/parachain/connection.go:25]: ditto: Just `API()` will suffice.

File: relayer/core/relay.go [hunk 3]
```diff
@@ -22,13 +22,16 @@ import (
 	"github.com/snowfork/polkadot-ethereum/relayer/chain/parachain"
 	"github.com/snowfork/polkadot-ethereum/relayer/chain/relaychain"
 	"github.com/snowfork/polkadot-ethereum/relayer/store"
+	"github.com/snowfork/polkadot-ethereum/relayer/workers/parachaincommitmentrelayer"
```
Comment (vgeddes) [inline relayer/core/relay.go:25]: Nit: The package name `parachaincommitmentrelayer` is a bit too long, and `commitment` is just an internal implementation detail. So can shorten the name to `parachainrelayer`.

File: relayer/core/relay.go [hunk 4]
```diff
@@ -22,13 +22,16 @@ import (
...
-	ethChain   chain.Chain
-	paraChain  chain.Chain
-	relayChain chain.Chain
-	database   *store.Database
+	ethChain                   chain.Chain
+	paraChain                  chain.Chain
+	relayChain                 chain.Chain
+	database                   *store.Database
+	direction                  Direction
+	parachainCommitmentRelayer *parachaincommitmentrelayer.Worker
```
Comment (vgeddes) [inline relayer/core/relay.go:34]: ditto: Can shorten the name to `parachainRelayer`.

File: relayer/core/relay.go [hunk 5]
```diff
@@ -44,12 +47,17 @@ type RelayConfig struct {
 	HeadersOnly bool      `mapstructure:"headers-only"`
 }
 
+type WorkerConfig struct {
+	ParachainCommitmentRelayer bool `mapstructure:"parachaincommitmentrrelayer"`
```
Comment (vgeddes) [inline relayer/core/relay.go:51]: ditto: Can shorten the name to parachainRelayer.

File: relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/sirupsen/logrus"
+
+	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
+	"github.com/snowfork/polkadot-ethereum/relayer/contracts/inbound"
+	"github.com/snowfork/polkadot-ethereum/relayer/substrate"
+)
+
+type EthereumChannelWriter struct {
```
Comment (vgeddes) [inline relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go:20]: Can we shorten the name of the file to `writer.go` and the name of this struct to `Writer`? Its already namespaced so there won't be any symbol clashes.

File: relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go [hunk 7]
```diff
@@ -14,26 +14,27 @@ import (
...
 
 	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/parachain"
 	chainTypes "github.com/snowfork/polkadot-ethereum/relayer/substrate"
 )
 
-type Listener struct {
-	config   *Config
-	conn     *Connection
+type ParachainCommitmentListener struct {
```
Comment (vgeddes) [inline relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go:21]: Ditto: can we shorten the name of this file and struct to `listener.go` and `Listener` respectively?

File: relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go [hunk 8]
```diff
@@ -14,26 +14,27 @@ import (
...
+	config   *parachain.Config
+	conn     *parachain.Connection
 	messages chan<- []chain.Message
 	log      *logrus.Entry
 }
 
-func NewListener(config *Config, conn *Connection, messages chan<- []chain.Message, log *logrus.Entry) *Listener {
-	return &Listener{
+func NewParachainCommitmentListener(config *parachain.Config, conn *parachain.Connection, messages chan<- []chain.Message, log *logrus.Entry) *ParachainCommitmentListener {
```
Comment (vgeddes) [inline relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go:28]: Lets break this method signature up along multiple lines for better readability.

**Review by vgeddes — COMMENTED** [review]
Looks good. :+1:

Just need to rework some naming and formatting for improved readability.

**Comment by musnit** [comment]
@vgeddes will do follow up PR. issues here: https://github.com/Snowfork/polkadot-ethereum/issues/359

File: relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	})
+
+	return nil
+}
+
+func (wr *EthereumChannelWriter) onDone(ctx context.Context) error {
+	wr.log.Info("Shutting down writer...")
+	// Avoid deadlock if a listener is still trying to send to a channel
+	for range wr.messages {
+		wr.log.Debug("Discarded message")
```
Comment (philipstanislaus) [inline relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go:67]: Is there a call missing on each message? We are just logging here?

File: relayer/workers/parachaincommitmentrelayer/main.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+	ethereumKp, err := secp256k1.NewKeypairFromString(ethereumConfig.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	parachainConn := parachain.NewConnection(parachainConfig.Endpoint, nil, log)
+	ethereumConn := ethereum.NewConnection(ethereumConfig.Endpoint, ethereumKp, log)
+	// channel for messages from substrate
+	var subMessages = make(chan []chain.Message, 1)
```
Comment (philipstanislaus) [inline relayer/workers/parachaincommitmentrelayer/main.go:43]: Why is this channel buffered?

**Review by philipstanislaus — COMMENTED** [review]
2 minor questions, LGTM otherwise 👍

File: relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go (hunk 9 above)
Comment (musnit) [reply relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go:67]: could be yeh - this code was just ported over from old code. will fix deadlock/restart logic in upcoming PR

File: relayer/workers/parachaincommitmentrelayer/main.go (hunk 10 above)
Comment (musnit) [reply relayer/workers/parachaincommitmentrelayer/main.go:43]: also ported from old code, likely not necessary - will address with other deadlock related fixes if needed

**Pull request merged**